libs/spec/Elf.3d

```diff
@@ -0,0 +1,460 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+/*
+
+  3d Specification of the 64-bit ELF format
+
+  ELF resources: https://man7.org/linux/man-pages/man5/elf.5.html
+                 https://github.com/torvalds/linux/blob/master/include/uapi/linux/elf.h
+                 https://www.youtube.com/watch?v=nC1U1LJQL8o
+
+  An ELF file consists of an ELF header, a program header table, and a section header table.
+
+  Both the tables are optional.
+
+  The ELF header starts at offset 0 in the file, and among other fields, contains the
+  offsets of and the number of entries in the program header table and the section header table.
+
+  These two tables then describe the rest of the file. Roughly:
+              -------------------------------
+             |                               |
+             |                  -------------|------------------
+             |                 |             |                  |
+             |                 |            \ /                \ /
+   ----------------------------------------------------------------------
+  |..|ph_offset|ph_num|..|sh_offset|sh_num|..|ph table|file data|sh table|
+   ----------------------------------------------------------------------
+
+  |<---------------ELF Header--------------->|
+
+  Where ph_table contains ph_num entries, sh_table contains sh_num entries,
+  and both describe sections in the rest of the file (i.e. in the file data part)
+
+  This specification validates the:
+  - ELF header
+  - Program Header table (that it contains ph_num entries,
+                          and points to valid parts of the rest of the file)
+  - Section Header table (that it contains sh_num entries,
+                          and points to valid parts of the rest of the file)
+
+  Though it is possible to do it in 3d, we do not validate individual sections/segments yet.
```

Comment:
The PR description says it resolves 834.  Is it really sufficient to resolve it if we don't validate individual sections?

Reply:
#834 is about building the parse from the .3d file as part of the build.

A second effort will be required to extend the basic ELF parser to a BPF specific ELF parser (that checks the sections etc).